src/identity/identity.service.ts

```diff
@@ -0,0 +1,136 @@
+    }
+  }
+
+  /**
+   * Check if the local login is enabled
+   * @throws {FunctionalityDisabledError} user login is disabled
+   */
+  checkIfLoginIsEnabled(): void {
+    if (!this.authConfig.email.enableLogin) {
+      this.logger.debug('User login is disabled.', 'checkIfLoginIsEnabled');
+      throw new FunctionalityDisabledError('User login is disabled.');
+    }
+  }
+
+  /**
+   * @async
+   * Create a new identity for internal auth
+   * @param {User} user - the user the identity should be added to
+   * @param {string} password - the password the identity should have
+   * @throws {FunctionalityDisabledError} user registration is disabled
+   * @return {Identity} the new local identity
+   */
+  async createLocalIdentity(user: User, password: string): Promise<Identity> {
+    this.checkIfRegistrationEnabled();
+    const identity = Identity.create(ProviderType.LOCAL);
+    identity.passwordHash = await hashPassword(password);
+    identity.user = user;
+    identity.providerName = '';
+    identity.syncSource = false;
+    return await this.identityRepository.save(identity);
+  }
+
+  /**
+   * @async
+   * Change the internal password of the specified the user
+   * @param {User} user - the user, which identity should be updated
+   * @param {string} newPassword - the new password
+   * @throws {FunctionalityDisabledError} user login is disabled
+   * @throws {NotInDBError} the specified user has no internal identity
+   * @return {Identity} the changed identity
+   */
+  async changeLocalPassword(
+    user: User,
+    newPassword: string,
+  ): Promise<Identity> {
+    this.checkIfLoginIsEnabled();
+    const internalIdentity: Identity | undefined = getIdentityFromUser(
+      user,
+      ProviderType.LOCAL,
+    );
+    if (internalIdentity === undefined) {
+      this.logger.debug(
+        'This user has no internal identity.',
+        'changeLocalPassword',
+      );
+      throw new NotInDBError('This user has no internal identity.');
+    }
+    internalIdentity.passwordHash = await hashPassword(newPassword);
+    return await this.identityRepository.save(internalIdentity);
+  }
+
+  /**
+   * @async
+   * Login the user with their username and password
+   * @param {User} user - the user to use
+   * @param {string} password - the password to use
+   * @throws {FunctionalityDisabledError} user login is disabled
+   * @throws {NotInDBError} the specified user can't be logged in
+   */
+  async loginWithLocalIdentity(user: User, password: string): Promise<void> {
+    this.checkIfLoginIsEnabled();
+    const internalIdentity: Identity | undefined = getIdentityFromUser(
+      user,
+      ProviderType.LOCAL,
+    );
+    if (internalIdentity === undefined) {
+      this.logger.debug(
+        `The user with the username ${user.userName} does not have a internal identity.`,
+        'loginWithLocalIdentity',
+      );
+      throw new NotInDBError(
+        'This username and password combination did not work.',
```

Comment:
Hmm, shouldn't the controller-layer create this user-facing error message?

Reply:
I think I fixed this…